onnx/defs/printer.cc

```diff
@@ -3,108 +3,211 @@
  */
 
 #include "onnx/defs/printer.h"
+#include <iomanip>
 #include "onnx/defs/tensor_proto_util.h"
 
 namespace ONNX_NAMESPACE {
 
-template <typename Collection>
-inline void print(std::ostream& os, const char* open, const char* separator, const char* close, Collection coll) {
-  const char* sep = "";
-  os << open;
-  for (auto& elt : coll) {
-    os << sep << elt;
-    sep = separator;
+using MetaDataProp = StringStringEntryProto;
+using MetaDataProps = google::protobuf::RepeatedPtrField<StringStringEntryProto>;
+
+class ProtoPrinter {
+ public:
+  ProtoPrinter(std::ostream& os) : output(os) {}
+
+  void print(const TensorShapeProto_Dimension& dim);
+
+  void print(const TensorShapeProto& shape);
+
+  void print(const TypeProto_Tensor& tensortype);
+
+  void print(const TypeProto& type);
+
+  void print(const TypeProto_Sequence& seqType);
+
+  void print(const TypeProto_Map& mapType);
+
+  void print(const TypeProto_Optional& optType);
+
+  void print(const TypeProto_SparseTensor& sparseType);
+
+  void print(const TensorProto& tensor);
+
+  void print(const ValueInfoProto& value_info);
+
+  void print(const ValueInfoList& vilist);
+
+  void print(const AttributeProto& attr);
+
+  void print(const AttrList& attrlist);
+
+  void print(const NodeProto& node);
+
+  void print(const NodeList& nodelist);
+
+  void print(const GraphProto& graph);
+
+  void print(const FunctionProto& fn);
+
+  void print(const ModelProto& model);
+
+  void print(const OperatorSetIdProto& opset);
+
+  void print(const OpsetIdList& opsets);
+
+  void print(const MetaDataProps& metadataprops) {
+    printSet("[", ", ", "]", metadataprops);
+  }
+
+  void print(const MetaDataProp& metadata) {
+    printQuoted(metadata.key());
+    output << ": ";
+    printQuoted(metadata.value());
   }
-  os << close;
-}
 
-std::ostream& operator<<(std::ostream& os, const TensorShapeProto_Dimension& dim) {
+ private:
+  template <typename T>
+  inline void print(T prim) {
+    output << prim;
+  }
+
+  void printQuoted(const std::string& str) {
+    output << "\"";
+    for (const char* p = str.c_str(); *p; ++p) {
+      if ((*p == '\\') || (*p == '"'))
+        output << '\\';
+      output << *p;
+    }
+    output << "\"";
+  }
+
+  template <typename T>
+  inline void printKeyValuePair(KeyWordMap::KeyWord key, const T& val, bool addsep = true) {
+    if (addsep)
+      output << "," << std::endl;
+    output << std::setw(indent_level) << ' ' << KeyWordMap::ToString(key) << ": ";
+    print(val);
+  }
+
+  inline void printKeyValuePair(KeyWordMap::KeyWord key, const std::string& val) {
+    output << "," << std::endl;
+    output << std::setw(indent_level) << ' ' << KeyWordMap::ToString(key) << ": ";
+    printQuoted(val);
```

Comment:
Maybe replace `printQuoted` with `print` and implement a `print` for `std::string`? Then, we can merge these  two `printKeyValuePair`s.

Reply:
That's how I started. But ran into C++ template-specialization issues (template-specialization is allowed for class-members only by some compilers, not all.) Furthermore, we use `print(str)` as well as `printQuoted(str)`, both are needed.